### DISCUSSION TITLE
Suggestion for a new InputModeAttribute attribute in Data Annotations

### DISCUSSION THREAD

**Discussion by Andres10121997** [dotnet/aspnetcore#58036] [Ideas]
Hello ASP.NET team,
I would like to propose the addition of a new Data Annotation attribute called `InputModeAttribute`, which would allow you to specify the HTML attribute `inputmode` in models. This functionality would improve the usability of forms in web applications, providing developers with an easy way to control keyboard behavior on mobile devices.

# Attribute Details
The `InputModeAttribute` would allow developers to define the input mode to be used in a specific field. This is especially useful in fields such as email, numbers, and searches.

# Proposed Enumeration
```
public class InputModeAttribute : ValidationAttribute
{
    public And InputModeen
    {
        @decimal,
        email,
        none,
        numeric,
        search,
        tel
        text,
        url
    }

    public InputModeEnum InputMode { get; }

    public InputModeAttribute(InputModeEnum inputMode)
    {
        InputMode = inputMode;
    }

    public string GetInputMode() => InputMode.ToString();
}
```

# Example of Use
Developers could use the new attribute in their models as follows:
```
public class MyModel
{
    [Required]
    [InputMode(InputModeAttribute.InputModeEnum.email)]
    public string Email { get; set; }

    [Required]
    [InputMode(InputModeAttribute.InputModeEnum.numeric)]
    public string Edad { get; set; }
}
```

# Proceeds
1. **Improved Usability:** It will facilitate data entry on mobile devices by displaying appropriate keyboards for each type of input.
2. **Simplicity:** Developers will be able to define keyboard behavior directly in their models, improving code clarity and maintainability.
3. **Consistency:** It will provide a uniform approach to handling `inputmode` throughout the application.

# Conclusion
I hope this suggestion will be considered for future versions of ASP.NET. I'm open to discussing more about this functionality and getting feedback. Thank you for your attention!

**Comment by martincostello** [comment]
Data Annotations live in the [dotnet/runtime](https://github.com/dotnet/runtime) repository - I suggest re-posting this there.

**Reply by Andres10121997** [reply]
Ok, thanks, I'll repost there.
Thanks.

**Reply by Andres10121997** [reply]
Hi martincostello. I'd like to briefly summarize the conversation I had in the `dotnet/runtime` repository regarding the suggestion I initially posted here:

First response from huoyaoyuan: The user pointed out that for numeric fields, the common practice is to define them as numeric types, rather than text strings. He also mentioned that validations like `[Url]` and `[EmailAddress]` already exist, and suggested that an attribute like `InputModeAttribute` would not be useful, since it cannot be changed dynamically at runtime.

My response: I appreciated the comment and explained that, while I recognize that using numeric types is preferable and that validations already exist, my proposal is focused on improving usability on mobile devices. The goal of `InputModeAttribute` is to make it easier for developers to specify the keyboard that should appear on specific fields, which improves the user experience, especially on mobile. I clarified that, while it is not dynamic, it can help standardize interaction and increase consistency of forms.

Second answer from huoyaoyuan: He commented that `ValidationAttribute` is more focused on data validation and not on HTML element manipulation, and suggested that these discussions be brought to the ASP.NET Core repository, as this type of functionality is more related to the frontend part. He also asked me about what HTML application model I was referring to, highlighting that Razor requires raw HTML and that Blazor uses different tags for different input modes.

My clarification: I replied stating that my proposal was primarily aimed at forms generated with Razor Pages or MVC, where models define the form fields. Also, I mentioned that I had already opened a discussion in the ASP.NET Core repository, but was told that `Data Annotations` belong in the `dotnet/runtime` repository, which was the reason I reposted there.

Final response from huoyaoyuan: Finally, he mentioned that although the final review and implementation of this type of functionality belongs to the `dotnet/runtime` repository, its use in application models like ASP.NET or languages ​​like C# depends on their own repositories. He commented that if there is no direct integration from the application models, it would not make sense to include the attribute in the BCL.

Here is the [link to the republishing](https://github.com/dotnet/runtime/discussions/108150) in the `dotnet/runtime` repository, where the discussion is taking place. Thank you for your time and initial guidance.
